src/main/java/com/mraof/minestuck/entity/animation/MobAnimationPhases.java

```diff
@@ -0,0 +1,141 @@
+package com.mraof.minestuck.entity.animation;
+
+import net.minecraft.world.entity.PathfinderMob;
+
+/**
+ * A supplement to MobAnimation that allows for certain code to be performed at the point where one phase of an animation ends and the other begins.
+ * Primarily intended for use with {@link com.mraof.minestuck.entity.ai.MobAnimationPhaseGoal}
+ */
+public class MobAnimationPhases
+{
+	private final int initiationStart;
+	private final int contactStart;
+	private final int recoveryStart;
+	private final int recoveryEnd;
+	
+	public MobAnimationPhases(int initiationStart, int contactStart, int recoveryStart, int recoveryEnd)
+	{
+		this.initiationStart = initiationStart;
+		this.contactStart = contactStart;
+		this.recoveryStart = recoveryStart;
+		this.recoveryEnd = recoveryEnd;
+	}
+	
+	public Phases getCurrentPhase(int time)
+	{
+		if(time < initiationStart)
+			return Phases.ANTICIPATION;
+		else if(time < contactStart)
+			return Phases.INITIATION;
+		else if(time < recoveryStart)
+			return Phases.CONTACT;
+		else if(time < recoveryEnd)
+			return Phases.RECOVERY;
+		else
+			return Phases.NEUTRAL;
+	}
+	
+	public int getInitiationStartTime()
+	{
+		return initiationStart;
+	}
+	
+	public int getContactStartTime()
+	{
+		return contactStart;
+	}
+	
+	public int getRecoveryStartTime()
+	{
+		return recoveryStart;
+	}
+	
+	/**
+	 * Equivalent to getting recoveryEnd value
+	 */
+	public int getTotalAnimationLength()
+	{
+		return recoveryEnd;
+	}
+	
+	/**
+	 * Is called every tick to check whether its time to transition to a new phase
+	 */
+	public <T extends PathfinderMob & MobAnimationPhases.Phases.Holder> void attemptPhaseChange(int time, T entity, MobAnimation animation)
+	{
+		if(time == getInitiationStartTime())
+			entity.setAnimationPhase(Phases.INITIATION, animation.getAction());
+		else if(time == getContactStartTime())
+			entity.setAnimationPhase(Phases.CONTACT, animation.getAction());
+		else if(time == getRecoveryStartTime())
+			entity.setAnimationPhase(Phases.RECOVERY, animation.getAction());
+		else if(time >= getTotalAnimationLength())
+			entity.setAnimationPhase(Phases.NEUTRAL, animation.getAction());
+	}
+	
+	/**
+	 * A list of all the phases, the transition to ANTICIPATION is not included in a MobAnimationPhases object because the transition time is equivalent to the start of when its called.
+	 * Contains the Holder interface
+	 */
+	public enum Phases
```

Comment:
it looks strongly related to attacks im not sure that has the right name for it

Reply:
Attacks are likely going to be the most common usage of them, but I was trying to account for other complex animations that may get added in the future. For example consort actions where they make a sound at a certain stage phase or perhaps particles to express emotions. I will consider the change!

Reply:
for that usage there are special keyframes that we can add the the animations . i remember having tried them but i dont remember if they work :D

Reply:
The keyframes would probably cover the majority of our needs, agreed! However there are some things server side I forgot about, such as the exploding consort. Do the keyframes only work clientside, and specifically with particles?

Reply:
We decided that keyframes would probably cover what we need. They can replace this system if we get them figured out